flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+        SavepointRestoreSettings.toConfiguration(savepointRestoreSettings, configuration);
+
+        final String expectedSuffix = "suffix";
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        null,
+                        classpathProvider.getTestJobClassName(),
+                        classpathProvider.getTestJobArgs(expectedSuffix),
+                        new Configuration());
+
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, configuration);
+
+        assertThat(jobGraph.getSavepointRestoreSettings(), is(savepointRestoreSettings));
+        assertThat(jobGraph.getJobID(), is(jobId));
+    }
+
+    @Test
+    public void testFailIfJobDirDoesNotHaveEntryClass() {
+        try {
+            PackagedProgramRetrieverImpl.create(
+                    classpathProvider.getUserDirectoryWithNoEntryClass(),
+                    classpathProvider.getTestJobClassName(),
+                    classpathProvider.getTestJobArgs("suffix"),
+                    new Configuration());
+            fail("This case should throw exception !");
+        } catch (FlinkException e) {
+            assertThat(
+                    ExceptionUtils.findThrowableWithMessage(
+                                    e,
+                                    String.format(
+                                            "Could not find the provided job class (%s) in the user lib directory.",
+                                            classpathProvider.getTestJobClassName()))
+                            .isPresent(),
+                    is(true));
+        }
+    }
+
+    // TODO: this test checks the same code path as
+    // testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName
+    // We should make it fail early if the class is not present on the system classpath
+    // Right now, the test is failing because no error is thrown
+    @Ignore
+    @Test(expected = FlinkException.class)
+    public void testEntryClassNotFoundOnSystemClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(
+                null, "NotExistingClass", new String[0], new Configuration());
+    }
+
+    @Test(expected = FlinkException.class)
+    public void testEntryClassNotFoundOnUserClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(
+                classpathProvider.getUserDirectoryWithNoEntryClass(),
+                classpathProvider.getTestJobClassName(),
+                new String[0],
+                new Configuration());
+    }
+
+    @Test(expected = FlinkException.class)
+    public void testTooManyEntryClassesOnSystemClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(null, null, new String[0], new Configuration());
+    }
+
+    @Test
+    public void testTooManyEntryClassesOnUserClasspath() throws FlinkException, IOException {
```

Comment:
So if 2 entry points are on the user CP we don't fail?

Reply:
No, it also fails if there are too many entrypoints on the classpath but no job class name is specified. But you're right. The test method's name is not specific enough. I updated the method name and additional tests. I realized that I wasn't precise here.